graphql/resolvers.go

```diff
@@ -244,6 +246,84 @@ func (r *mutationResolver) SpawnHost(ctx context.Context, spawnHostInput *SpawnH
 	return &apiHost, nil
 }
 
+func (r *mutationResolver) UpdateSpawnHostStatus(ctx context.Context, hostID string, action string) (*restModel.APIHost, error) {
+	host, err := host.FindOneByIdOrTag(hostID)
+	if err != nil {
+		return nil, ResourceNotFound.Send(ctx, fmt.Sprintf("Error finding host by id: %s", err))
+	}
+	usr := route.MustHaveUser(ctx)
+	env := evergreen.GetEnvironment()
+	queue := env.RemoteQueue()
+
+	if usr.Username() != host.StartedBy {
+		if !usr.HasPermission(gimlet.PermissionOpts{
+			Resource:      host.Distro.Id,
+			ResourceType:  evergreen.DistroResourceType,
+			Permission:    evergreen.PermissionHosts,
+			RequiredLevel: evergreen.HostsEdit.Value,
+		}) {
+			return nil, Forbidden.Send(ctx, "You are not authorized to modify this host")
+		}
+	}
+	var (
+		hostTerminate = "terminate"
+		hostStop      = "stop"
+		hostStart     = "start"
+	)
+	switch action {
+	case hostStart:
+		if host.Status == evergreen.HostStarting || host.Status == evergreen.HostRunning {
+			return nil, InternalServerError.Send(ctx, fmt.Sprintf("Host %v is already starting or running", hostID))
+		}
+		// Start the host
+		ts := utility.RoundPartOfMinute(1).Format(units.TSFormat)
+		startJob := units.NewSpawnhostStartJob(host, usr.Id, ts)
+		if err = queue.Put(ctx, startJob); err != nil {
+			return nil, InternalServerError.Send(ctx, fmt.Sprintf("Error while trying to start a host : %s", err))
+		}
+		apiHost := restModel.APIHost{}
+		err = apiHost.BuildFromService(host)
+		if err != nil {
+			return nil, InternalServerError.Send(ctx, fmt.Sprintf("Error building apiHost from service: %s", err))
+		}
+		return &apiHost, nil
+	case hostStop:
+		if host.Status == evergreen.HostStopped {
+			return nil, InternalServerError.Send(ctx, fmt.Sprintf("Host %v is already Stopped", hostID))
+		}
+
+		// Stop the host
+		ts := utility.RoundPartOfMinute(1).Format(units.TSFormat)
+		stopJob := units.NewSpawnhostStopJob(host, usr.Id, ts)
+		if err = queue.Put(ctx, stopJob); err != nil {
+			return nil, InternalServerError.Send(ctx, fmt.Sprintf("Error while stopping a host : %s", err))
+		}
+		apiHost := restModel.APIHost{}
+		err = apiHost.BuildFromService(host)
+		if err != nil {
+			return nil, InternalServerError.Send(ctx, fmt.Sprintf("Error building apiHost from service: %s", err))
+		}
+		return &apiHost, nil
+	case hostTerminate:
+		if host.Status == evergreen.HostTerminated {
+			return nil, InternalServerError.Send(ctx, fmt.Sprintf("Host %v is already terminated", hostID))
+		}
+		err = cloud.TerminateSpawnHost(ctx, env, host, usr.Id, fmt.Sprintf("terminated via UI by %s", usr.Username()))
+		if err != nil {
+			return nil, InternalServerError.Send(ctx, fmt.Sprintf("Error while terminating host : %s", err))
+		}
+		apiHost := restModel.APIHost{}
+		err = apiHost.BuildFromService(host)
+		if err != nil {
+			return nil, InternalServerError.Send(ctx, fmt.Sprintf("Error building apiHost from service: %s", err))
+		}
+		return &apiHost, nil
```

Comment:
this kind of logic should be shareable between the graphql and the rest APIs. mashing all the business logic in with the resolvers will result in same mess as in the rest API handlers are currently. can these be moved to a shared folder between the two packages that can be used agnostic of whether the api is gql or rest?